metals/src/main/scala/scala/meta/internal/metals/callHierarchy/CallHierarchyProvider.scala

```diff
@@ -0,0 +1,619 @@
+
+                Future
+                  .sequence(results.map {
+                    case FindIncomingCallsResult(
+                          occurence @ SymbolOccurrence(_, symbol, _),
+                          range,
+                          ranges,
+                        ) if !containsDuplicates(info.visited) =>
+                      buildCallHierarchyItem(
+                        source,
+                        doc,
+                        occurence,
+                        range,
+                        info.visited :+ symbol,
+                        token,
+                      )
+                        .mapOptionInside(chi =>
+                          new CallHierarchyIncomingCall(
+                            chi,
+                            ranges.asJava,
+                          )
+                        )
+                    case _ => Future.successful(None)
+                  })
+                  .map(_.flatten)
+              case None =>
+                Future.successful(Nil)
+            }
+          })
+      )
+    } yield (result.flatten)
+  }
+
+  private def findDefinitionOccurence(
+      source: AbsolutePath,
+      symbol: String,
+  )(implicit
+      ec: ExecutionContext
+  ): Future[Option[(SymbolOccurrence, AbsolutePath, TextDocument)]] =
+    (if (!symbol.isLocal)
+       references
+         .pathsMightContainSymbol(
+           source,
+           Set(symbol),
+         )
+     else Future.successful(Set(source)))
+      .map(paths =>
+        paths.view
+          .map(source =>
+            for {
+              doc <- semanticdbs.textDocument(source).documentIncludingStale
+              occ <- doc.occurrences.find(occ =>
+                occ.symbol == symbol && occ.role.isDefinition && doc.symbols
+                  .exists(symInfo => symInfo.symbol == symbol)
+              )
+            } yield (occ, source, doc)
+          )
+          .find(_.isDefined)
+          .flatten
+      )
+
+  private def findOutgoingCalls(
+      source: AbsolutePath,
+      doc: TextDocument,
+      root: Tree,
+  )(implicit ec: ExecutionContext): Future[List[
+    FindOutgoingCallsResult
+  ]] = {
+    val realRoot = findRealRoot(root)
+
+    def search(
+        tree: Tree
+    ): Future[
+      List[FindOutgoingCallsResult]
+    ] =
+      tree match {
+        case name: Name if !isTypeDeclaration(name) =>
+          (for {
+            definitionInfo <- Future
+              .sequence(
+                definition
+                  .positionOccurrences(source, name.pos.toLSP.getEnd, doc)
+                  .flatMap(rso =>
+                    rso.occurrence
+                      .map(occ => findDefinitionOccurence(source, occ.symbol))
+                  )
+              )
+              .map(
+                _.flatten
+                  .sortBy(
+                    _._1.symbol.length * -1
+                  )
+                  .headOption // The most specific occurrence is the longest
```

Comment:
I find this pretty hard to grok without spending time and digging into it. Maybe refactoring this either into a for comprehension or something else would be easier to read.